src/main.rs

```diff
@@ -0,0 +1,270 @@
+extern crate clap;
+extern crate crypto;
+#[macro_use]
+extern crate nickel;
+extern crate nickel_mustache;
+extern crate rustc_serialize;
+extern crate rand;
+extern crate postgres;
+
+
+use std::path::Path;
+use std::ascii::AsciiExt;
+use std::collections::HashMap;
+use std::error::Error;
+use std::fs::File;
+use std::fmt;
+use std::io::prelude::*;
+use rand::os::OsRng;
+use rand::Rng;
+
+use nickel::status::StatusCode;
+use nickel::{Nickel, HttpRouter, FormBody};
+use nickel_mustache::Render;
+
+use clap::{App, Arg};
+
+use crypto::bcrypt;
+
+use postgres::{Connection, TlsMode};
+
+
+fn main() {
+    let matches = App::new("synapse-password-reset")
+        .version("v0.0.1")
+        .author("Euan Kemp <euank@euank.com>")
+        .args(&[Arg::with_name("token-dir")
+                    .help("sets the database directory to use")
+                    .takes_value(true)
+                    .short("t")
+                    .long("token-dir")
+                    .required(true),
+                Arg::with_name("pepper")
+                    .help("sets the hash pepper (e.g. from your synapse config)")
+                    .takes_value(true)
+                    .short("p")
+                    .long("pepper")
+                    .required(true),
+                Arg::with_name("db")
+                    .help("sets the postgres db to connect to (including username,pass)")
+                    .takes_value(true)
+                    .short("d")
+                    .long("db")
+                    .required(true)])
+        .get_matches();
+
+    let mut server = Nickel::new();
+
+    server.get("/",
+               middleware!{ |_, res|
+        let data: HashMap<String, String> = HashMap::new();
+        return Render::render(res, "public/index.tpl", &data);
+    });
+
+    server.post("/",
+                middleware! {|req, mut res|
+
+        let token_dir = matches.value_of("token-dir").unwrap();
+        let pepper = matches.value_of("pepper").unwrap();
+        let db = matches.value_of("db").unwrap();
```

Comment:
What does `unwrap` do here — does it panic, or does the `middleware!` macro take care of it and return a 500?

I know it's theoretically impossible to get to a point where this panics (assuming `.required(true)` does the right thing in clap), but to me it makes more sense for this to be directly in `main()` right after `App::get_matches()` instead of in the closure.